target/linux/rtl838x/files-5.4/drivers/net/dsa/rtl838x_sw.c

```diff
@@ -0,0 +1,2152 @@
+		break;
+	case BR_STATE_FORWARDING: /* 3*/
+		port_state[index] |= (3 << bit);
+	default:
+		break;
+	}
+
+	if (priv->family_id == RTL8380_FAMILY_ID) {
+		cmd = 1 << 15 /* Execute cmd */
+			| 0 << 14 /* Write */
+			| 2 << 12 /* Table type 0b10 */
+			| (msti & 0xfff);
+	} else {
+		cmd = 1 << 16 /* Execute cmd */
+			| 1 << 15 /* Write */
+			| 5 << 12 /* Table type 0b101 */
+			| (msti & 0xfff);
+	}
+	for (i = 0; i < n; i++)
+		sw_w32(port_state[i], priv->r->tbl_access_data_0(i));
+	priv->r->exec_tbl0_cmd(cmd);
+
+	mutex_unlock(&priv->reg_mutex);
+}
+
+static int rtl838x_port_mirror_add(struct dsa_switch *ds, int port,
+				   struct dsa_mall_mirror_tc_entry *mirror,
+				   bool ingress)
+{
+	/* We support 4 mirror groups, one destination port per group */
+	int group;
+	struct rtl838x_switch_priv *priv = ds->priv;
+
+	pr_info("In %s\n", __func__);
+
+	for (group = 0; group < 4; group++) {
+		if (priv->mirror_group_ports[group] == mirror->to_local_port)
+			break;
+	}
+	if (group >= 4) {
+		for (group = 0; group < 4; group++) {
+			if (priv->mirror_group_ports[group] < 0)
+				break;
+		}
+	}
+
+	if (group >= 4)
+		return -ENOSPC;
+
+	pr_debug("Using group %d\n", group);
+	mutex_lock(&priv->reg_mutex);
+	/* Enable mirroring to port across VLANs (bit 11) */
+	sw_w32(1 << 11 | (mirror->to_local_port << 4) | 1, RTL838X_MIR_CTRL(group));
+
+	if (ingress && (sw_r32(RTL838X_MIR_SPM_CTRL(group)) & (1 << port))) {
+		mutex_unlock(&priv->reg_mutex);
+		return -EEXIST;
+	}
+	if ( (!ingress) && (sw_r32(RTL838X_MIR_DPM_CTRL(group)) & (1 << port))) {
+		mutex_unlock(&priv->reg_mutex);
+		return -EEXIST;
+	}
+	if (ingress)
+		sw_w32_mask(0, 1 << port, RTL838X_MIR_SPM_CTRL(group));
+	else
+		sw_w32_mask(0, 1 << port, RTL838X_MIR_DPM_CTRL(group));
+
+	priv->mirror_group_ports[group] = mirror->to_local_port;
+	mutex_unlock(&priv->reg_mutex);
+	return 0;
+}
+
+static void rtl838x_port_mirror_del(struct dsa_switch *ds, int port,
+				    struct dsa_mall_mirror_tc_entry *mirror)
+{
+	int group = 0;
+	struct rtl838x_switch_priv *priv = ds->priv;
+
+	pr_info("In %s\n", __func__);
+	for (group = 0; group < 4; group++) {
+		if (priv->mirror_group_ports[group] == mirror->to_local_port)
+			break;
+	}
+	if (group >= 4)
+		return;
+
+	mutex_lock(&priv->reg_mutex);
+	if (mirror->ingress) {
+		/* Ingress, clear source port matrix */
+		sw_w32_mask(1 << port, 0, RTL838X_MIR_SPM_CTRL(group));
+	} else {
+		/* Egress, clear destination port matrix */
+		sw_w32_mask(1 << port, 0, RTL838X_MIR_DPM_CTRL(group));
+	}
+
+	if (!(sw_r32(RTL838X_MIR_DPM_CTRL(group)) || sw_r32(RTL838X_MIR_DPM_CTRL(group)))) {
+		priv->mirror_group_ports[group] = -1;
+		sw_w32(0, RTL838X_MIR_CTRL(group));
+	}
+
+	mutex_unlock(&priv->reg_mutex);
+}
+
+
+void rtl838x_vlan_profile_dump(int index)
+{
+	u32 profile;
+
+	if (index < 0 || index > 7)
+		return;
+
+	profile = sw_r32(RTL838X_VLAN_PROFILE(index));
+
+	pr_info("VLAN %d: L2 learning: %d, L2 Unknown MultiCast Field %x, "
```

Comment:
WARNING: quoted string split across lines

Reply:
Will be fixed.